### ISSUE TITLE
1.1.395 no longer respects strictParameterNoneValue=false in dataclass

### ISSUE THREAD

**Issue by XeCycle** [microsoft/pyright#9990]
**Describe the bug**
This is very similar to #9982 but involves dataclasses, so it's a bit different, and in my case I've already set `strictParameterNoneValue = false`.

**Code or Screenshots**

```python
from dataclasses import dataclass


@dataclass
class A:
    x: str = None


a = A()
```

```
 $ cat pyproject.toml 
[tool.pyright]
strictParameterNoneValue = false
 $ pyright --version
pyright 1.1.395
 $ pyright dc.py
dc.py
  dc.py:6:14 - error: Type "None" is not assignable to declared type "str"
    "None" is not assignable to "str" (reportAssignmentType)
  dc.py:9:5 - error: Argument of type "None" cannot be assigned to parameter "x" of type "str" in function "__init__"
    "None" is not assignable to "str" (reportArgumentType)
2 errors, 0 warnings, 0 informations

 $ pyrve/bin/pyright --version
pyright 1.1.394
 $ pyrve/bin/pyright dc.py    
dc.py
  dc.py:6:14 - error: Type "None" is not assignable to declared type "str"
    "None" is not assignable to "str" (reportAssignmentType)
1 error, 0 warnings, 0 informations
```

Let's assume `class A` is from an upstream library, and ignore the diagnostic on line 6; and now 1.1.395 rejects my code on line 9.

My actual code uses lightkube, which uses many dataclasses with such type hints.

**Comment by erictraut** [comment]
`strictParameterNoneValue` does not apply to dataclass field definitions, so it shouldn't have any effect here. The only thing that `strictParameterNoneValue` controls is whether a parameter in a `def` statement should be implicitly treated as `Optional` if it is assigned a default value of `None`. That doesn't apply in this case.

Prior to 1.1.395, pyright was inconsistent in how it applied default argument values to a call. If the corresponding parameter type was generic (and therefore had a type variable that needed to be solved), it attempted to assign the default argument value to the parameter and emitted an error if it wasn't compatible. If the parameter type was not generic, it didn't attempt to assign the default argument value to the parameter and therefore never emitted an error in this case.

I'm sympathetic to the fact that you are now seeing an error where you didn't previously — and that the source of the error is in a library that isn't directly under your control, but I think pyright's new behavior is more consistent and correct from a type checking standpoint. 

I recommend that you file a bug with the library maintainer and try to get them to fix the upstream problem.

In the meantime, you can use `# pyright: ignore` or `# type: ignore` comments to work around the problem in your code.

**Comment by XeCycle** [comment]
So we don't treat that `class A` as having a (generated) `def __init__(self, x: str = None)`?  If it was checked this way then it would have been controlled by `strictParameterNoneValue`.
